ndk/src/system_properties.rs

```diff
@@ -0,0 +1,326 @@
+///
+/// # Deprecation
+/// Deprecated since Android O (API level 26), use [`Property::find()`] with
+/// [`Property::read_callback()`] instead which does not have a limit on `value` nor `name` length.
+#[doc(alias = "__system_property_get")]
+pub fn get_raw(name: &CStr) -> Result<CString, GetRawError> {
+    process_owned(|value| unsafe { ffi::__system_property_get(name.as_ptr(), value) })
+}
+
+/// Possible failures returned by [`get()`].
+#[allow(missing_debug_implementations)] // Our MSRV is too low for derive(Debug) to emit bounds on T.
+#[derive(Error)]
+pub enum GetError<T> {
+    #[error("Property is missing or empty")]
+    MissingOrEmpty,
+    #[error(transparent)]
+    NulError(#[from] FromBytesWithNulError),
+    #[error("Property does not contain valid UTF-8")]
+    Utf8Error(#[from] Utf8Error),
+    #[error(transparent)]
+    Io(#[from] Error),
+    #[error(transparent)]
+    ParseError(T),
+}
+
+/// Internal helper to deduplicate the implementation between [`get()`] and [`Property::read()`].
+fn process_parse<T: FromStr>(get: impl FnOnce(*mut c_char) -> i32) -> Result<T, GetError<T::Err>> {
+    let mut value = [0u8; ffi::PROP_VALUE_MAX as usize];
+    let ret = get(value.as_mut_ptr());
+    match ret {
+        0 => Err(GetError::MissingOrEmpty),
+        ..=-1 => Err(Error::from_raw_os_error(-ret).into()),
+        1.. => {
+            let c_str = CStr::from_bytes_with_nul(&value[..ret as usize + 1])?;
+            c_str.to_str()?.parse().map_err(GetError::ParseError)
+        }
+    }
+}
+
+/// Returns the property value as a [`FromStr`]-parsed type from a source string of at most 92
+/// ([`ffi::PROP_VALUE_MAX`]) characters, including NUL terminator.
+///
+/// # Implementation details
+/// This is implemented without any up-front allocations like [`get_raw()`], but requires a trip
+/// through [`CStr`] and [`str`] (for calling [`FromStr::from_str()`]) meaning the resulting
+/// string has to be compliant with [`CStr`] ([no interior NULs]) and [`str`] ([valid UTF-8]).  In
+/// other words, parsing into a [`String`] will never contain interior NULs (and it is unknown and
+/// unlikely whether the property API allows for this).
+///
+/// [no interior NULs]: GetError::NulError
+/// [valid UTF-8]: GetError::Utf8Error
+///
+/// # Deprecation
+/// Deprecated since Android O (API level 26), use [`Property::find()`] with
+/// [`Property::read_callback()`] instead which does not have a limit on `value` nor `name` length.
+#[doc(alias = "__system_property_get")]
+pub fn get<T: FromStr>(name: &CStr) -> Result<T, GetError<T::Err>> {
+    process_parse(|value| unsafe { ffi::__system_property_get(name.as_ptr(), value) })
+}
+
+/// Sets system property `name` to `value`, creating the system property if it doesn't already
+/// exist.
+#[doc(alias = "__system_property_set")]
+pub fn set(name: &CStr, value: &CStr) -> std::io::Result<()> {
+    let ret = unsafe { ffi::__system_property_set(name.as_ptr(), value.as_ptr()) };
```

Comment:
If I check the source for `__system_property_set`, this is not allowed to exceed both `PROP_VALUE_MAX` and `PROP_NAME_MAX` lens. Probably should specify that in the docs and maybe assert as well.

If we don't assert, `__system_property_set` jjust returns `-1` which is cascaded to a generic error

Reply:
Yeah with `set()` this `last_os_error()` is actually [`13` which is `EACCESS`](https://github.com/torvalds/linux/blob/d082ecbc71e9e0bf49883ee4afd435a77a5101b6/include/uapi/asm-generic/errno-base.h#L17) (`Permission Denied`, not `EPERM` which is `Operation not permitted`).

Reply:
I've taken a dependency on `nix` just to clear the error in this very infrequently used path. Sensible or overdone?

Reply:
Since the latest `nix 0.29` is also requiring an MSRV bump to `1.89` and this function is very unlikely to be used by most `ndk` consumers, I've instead reverted it and created a wrapper type over `io::Error` specifically for `set()`, documenting and format-printing that its value may be stale/bogus.